### ISSUE TITLE
Performance degradation over time

### ISSUE THREAD

**Issue by vaibhavinbayarea** [codekitchen/logstash-input-kinesis#11]
Hi,

I'm using logstash-input-kinesis-1.4.3 and running into some performance degradation over time. On a fresh EC2 instance, fresh install performance is decent (1200-1400 eps) for the size of the instance. But over time (few hours), this comes down to 90-200 eps. 

Looking at Kinesis stats, I see the gradual decline, which never recovers. If I kill logstash and start again, the performance results are on the same lower range. Re-installing the plugin gives a momentary boost to 600 eps, but within few minutes it comes down to 90-200 eps.

The instance being used is t2.medium. I have enough cpu credits to warrant out throttling on aws end.

There is no heavy loading on cpu or memory pressure

```
top - 23:44:21 up 1 day,  1:16,  1 user,  load average: 0.10, 0.09, 0.12
Tasks: 106 total,   2 running, 104 sleeping,   0 stopped,   0 zombie
%Cpu(s):  7.2 us,  0.0 sy,  0.0 ni, 92.7 id,  0.0 wa,  0.0 hi,  0.2 si,  0.0 st
KiB Mem:   4046872 total,  1834504 used,  2212368 free,    99380 buffers
KiB Swap:        0 total,        0 used,        0 free.  1210504 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                       
 6967 ubuntu    20   0 3560200 375684  15924 S  15.3  9.3   1:58.98 java                                                                                                          
    1 root      20   0   33508   2844   1492 S   0.0  0.1   0:01.22 init                                                                                                          
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.00 kthreadd                                                                                                      
    3 root      20   0       0      0      0 S   0.0  0.0   0:02.25 ksoftirqd/0
```

```
$ free -m
 total       used       free     shared    buffers     cached
Mem:          3952       1791       2160          0         97       1182
-/+ buffers/cache:        511       3440
Swap:            0          0          0
```

```
/usr/bin/java -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -Djava.awt.headless=true -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -XX:+HeapDumpOnOutOfMemoryError -Xmx1g -Xss2048k -Djffi.boot.library.path=/opt/logstash/vendor/jruby/lib/jni -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -Djava.awt.headless=true -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/opt/logstash/heapdump.hprof -Xbootclasspath/a:/opt/logstash/vendor/jruby/lib/jruby.jar -classpath : -Djruby.home=/opt/logstash/vendor/jruby -Djruby.lib=/opt/logstash/vendor/jruby/lib -Djruby.script=jruby -Djruby.shell=/bin/sh org.jruby.Main --1.9 /opt/logstash/lib/bootstrap/environment.rb logstash/runner.rb agent -f /etc/logstash/conf.d -w 2 -b 500
```

I'm trying to profile the plugin if its getting stuck on specific call etc but haven't made much headway. Cant get logstash to print my debug statements :(

Any pointers or tips to debug this further would be helpful!

Thanks.

**Comment by vaibhavinbayarea** [comment]
Turned on INFO logging on KCL and I see the following:

```
Jun 23, 2016 2:00:03 AM com.amazonaws.services.kinesis.leases.impl.LeaseCoordinator <init>
INFO: With failover time 10000ms and epsilon 25ms, LeaseCoordinator will renew leases every 3308ms and take leases every 20050ms
Jun 23, 2016 2:00:03 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker initialize
INFO: Initialization attempt 1
Jun 23, 2016 2:00:03 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker initialize
INFO: Initializing LeaseCoordinator
Jun 23, 2016 2:00:03 AM com.amazonaws.services.kinesis.leases.impl.LeaseManager createLeaseTableIfNotExists
INFO: Table logstash already exists.
Jun 23, 2016 2:00:03 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker initialize
INFO: Syncing Kinesis shard info
Jun 23, 2016 2:00:03 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker initialize
INFO: Starting LeaseCoordinator
Jun 23, 2016 2:00:13 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker run
INFO: Initialization complete. Starting worker loop.
Jun 23, 2016 2:00:23 AM com.amazonaws.services.kinesis.leases.impl.LeaseTaker computeLeasesToTake
INFO: Worker 95b95dcd-6de2-4bca-8876-8aa38d84155a needed 1 leases but none were expired, so it will steal lease shardId-000000000000 from 6ccec748-0100-4cdb-9ecf-13b2ea02b78d
Jun 23, 2016 2:00:23 AM com.amazonaws.services.kinesis.leases.impl.LeaseTaker computeLeasesToTake
INFO: Worker 95b95dcd-6de2-4bca-8876-8aa38d84155a saw 2 total leases, 0 available leases, 2 workers. Target is 1 leases, I have 0 leases, I will take 1 leases
Jun 23, 2016 2:00:23 AM com.amazonaws.services.kinesis.leases.impl.LeaseTaker takeLeases
INFO: Worker 95b95dcd-6de2-4bca-8876-8aa38d84155a successfully took 1 leases: shardId-000000000000
Jun 23, 2016 2:00:24 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker infoForce
INFO: Created new shardConsumer for : ShardInfo [shardId=shardId-000000000000, concurrencyToken=565182a7-4314-4752-ad9a-c6c94e0c0480, parentShardIds=[]]
Jun 23, 2016 2:00:24 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.BlockOnParentShardTask call
INFO: No need to block on parents [] of shard shardId-000000000000
Jun 23, 2016 2:00:25 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.KinesisDataFetcher initialize
INFO: Initializing shard shardId-000000000000 with 49562686177254807473188589338490271906772918110666096642
rate: 49.6
rate: 49.631982234148275
rate: 53.6591865785776
rate: 55.365503807973816
rate: 56.935391444465594
Jun 23, 2016 2:01:03 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker info
INFO: Current stream shard assignments: shardId-000000000000
Jun 23, 2016 2:01:03 AM com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker info
INFO: Sleeping ...
rate: 58.33178444479308
rate: 61.66339101086118
```

rate is events per second being handled. Not sure why the sleep is there...

**Comment by codekitchen** [comment]
Are you seeing actual lag in events making it through the pipeline? This looks like the consumer is processing fewer messages simply because it's all caught up and there's fewer messages to process.

That sleep is from the KCL, https://github.com/awslabs/amazon-kinesis-client/blob/master/src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java#L359 , it sleeps when there are no messages to process to avoid hitting the kinesis API too frequently.

**Comment by vaibhavinbayarea** [comment]
Thanks for the response. After more debugging last night I figured out the plugin, logstash are behaving as expected. The issue is elasticsearch on the other end was throttling indexing as it couldn't handle all the data. I changed the replica size to 0, increased the refresh interval to 15s, in addition to some other tunable, I was able to sustain high rate of event processing (3500 eps). 

The issue can be closed as it seems to be my setup issue rather than the plugin. Apologies for the trouble.
